File: testdata/fixtures/python/eval-detected.py
# eval() parses its argument as Python code, so a caller-chosen string
# becomes code. Flag eval() of anything but a constant; a string literal
# cannot carry input.
from flask import request


def from_query():
    user = request.args.get("expr")
    # ruleid: eval-detected
    eval(user)
    # ruleid: eval-detected
    eval("1 + " + user)
    # ruleid: eval-detected
    result = eval(compile(user, "<expr>", "eval"))


def safe_literal():
    # ok: eval-detected
    eval("1 + 1")
    # ok: eval-detected
    eval("sorted([3, 1, 2])")
    # ok: eval-detected
    eval("")

File: testdata/fixtures/javascript/eval-detected.js
function runUser(code) {
  // ruleid: eval-detected
  eval(code);
}

function runAgain(input) {
  // ruleid: eval-detected
  eval(input);
}

function runIndirect(fn, code) {
  // ruleid: eval-detected
  fn(eval(code));
}

function runStatic() {
  // ok: eval-detected
  eval('2 + 2');
}

function runConst() {
  // ok: eval-detected
  eval('static');
}

function plain(code) {
  // ok: eval-detected
  console.log(code);
}
